src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,377 @@
+            const int NameTypeOffset = ServerNameLengthOffset + sizeof(ushort);
+            const int HostNameStructOffset = NameTypeOffset + sizeof(NameType);
+
+            if (serverName.Length < HostNameStructOffset)
+            {
+                invalid = true;
+                return null;
+            }
+
+            int hostNameStructLength = BinaryPrimitives.ReadUInt16BigEndian(serverName) - sizeof(NameType);
+            NameType nameType = (NameType)serverName[NameTypeOffset];
+            ReadOnlySpan<byte> hostNameStruct = serverName.Slice(HostNameStructOffset);
+
+            if (hostNameStructLength != hostNameStruct.Length || nameType != NameType.HostName)
+            {
+                invalid = true;
+                return null;
+            }
+
+            return GetSniFromHostNameStruct(hostNameStruct, out invalid);
+        }
+
+        private static string GetSniFromHostNameStruct(ReadOnlySpan<byte> hostNameStruct, out bool invalid)
+        {
+            // https://tools.ietf.org/html/rfc3546#section-3.1
+            // HostName is an opaque type (length of sufficient size for max data length is prepended)
+            const int HostNameLengthOffset = 0;
+            const int HostNameOffset = HostNameLengthOffset + sizeof(ushort);
+
+            int hostNameLength = BinaryPrimitives.ReadUInt16BigEndian(hostNameStruct);
+            ReadOnlySpan<byte> hostName = hostNameStruct.Slice(HostNameOffset);
+            if (hostNameLength != hostName.Length)
+            {
+                invalid = true;
+                return null;
+            }
+
+            invalid = false;
+            return DecodeString(hostName);
+        }
+
+        private static string DecodeString(ReadOnlySpan<byte> bytes)
+        {
+            // https://tools.ietf.org/html/rfc3546#section-3.1
+            // Per spec:
+            //   If the hostname labels contain only US-ASCII characters, then the
+            //   client MUST ensure that labels are separated only by the byte 0x2E,
+            //   representing the dot character U+002E (requirement 1 in section 3.1
+            //   of [IDNA] notwithstanding). If the server needs to match the HostName
+            //   against names that contain non-US-ASCII characters, it MUST perform
+            //   the conversion operation described in section 4 of [IDNA], treating
+            //   the HostName as a "query string" (i.e. the AllowUnassigned flag MUST
+            //   be set). Note that IDNA allows labels to be separated by any of the
+            //   Unicode characters U+002E, U+3002, U+FF0E, and U+FF61, therefore
+            //   servers MUST accept any of these characters as a label separator.  If
+            //   the server only needs to match the HostName against names containing
+            //   exclusively ASCII characters, it MUST compare ASCII names case-
+            //   insensitively.
+
+            string idnEncodedString = Encoding.UTF8.GetString(bytes);
```

Comment:
Treat UTF-8 decoding failures as error. See constructor on `Utf8Encoding` for options to enable these checks.